ksqldb-udf/src/main/java/io/confluent/ksql/schema/ksql/SqlArgument.java

```diff
@@ -15,31 +15,48 @@
 
 package io.confluent.ksql.schema.ksql;
 
+import io.confluent.ksql.schema.ksql.types.SqlLambda;
 import io.confluent.ksql.schema.ksql.types.SqlType;
 import java.util.Objects;
 
 /**
- * A wrapper class to bundle SqlTypes for UDF functions.
+ * A wrapper class to bundle SqlTypes and SqlLambdas for UDF functions that contain
+ * lambdas as an argument. This class allows us to properly find the matching UDF and
+ * resolve the return type for lambda UDFs based on the given sqlLambda.
  */
 public class SqlArgument {
 
   private final SqlType sqlType;
+  private final SqlLambda sqlLambda;
 
-  public SqlArgument(final SqlType type) {
+  public SqlArgument(final SqlType type, final SqlLambda lambda) {
```

Comment:
We should change the sqlType and sqlLambda variables to be Optionals so we don't need to do null checks where SqlArgument is used

Reply:
Now that I'm thinking about this more, I think we should actually have SqlArgument be a base class and two classes extend it SqlArgumentType (which has a SqlType) and SqlArgumentLambda(which has a SqlLambda). There are certain portions of the code such as the UDAF's where we now need to do some unnecessary null checks (they should only be supporting SqlArgumentType). The GenericsUtil can still take in SqlArgument and we can use `instance of` checks instead of null checks/optional checks when deciding what behavior to use.

Reply:
One idea for the optional is to have the getter functions like this
```
  public SqlType getSqlTypeOrThrow() {
    if (sqlType.isPresent()) {
      return sqlType.get();
    }
    if (sqlLambda.isPresent()) {
      throw new RuntimeException("Was expecting sql type as an argument");
    }
    return null;
  }

  public Optional<SqlLambda> getSqlLambda() {
    return sqlLambda;
  }

  public SqlLambda getSqlLambdaOrThrow() {
    if (sqlLambda.isPresent()) {
      return sqlLambda.get();
    }
    throw new RuntimeException("Was expecting sql lambda as an argument");
  }
```

We return null from `getSqlTypeOrThrow` if both the type and lambda are missing because this is when the function argument is `NULL`

Reply:
+1 this makes the contract way clearer